File: src/aizk/graph/session_tier.py
from loguru import logger
from sqlalchemy import func, update

from ..background.jobs.projection import enqueue_pending
from ..config import settings
from ..extract.ingest import TextSource, ingest_texts
from ..provenance import CaptureContext
from ..store import SessionItem
from ..store.identity import User
from ..types import Scopes


async def due_working_items(scopes: Scopes) -> list[SessionItem]:
    """The aged and overflow working items in one exact scope set, decided in the database."""
    async with User.system(scopes) as session:
        result = await session.exec(
            SessionItem.due_for_promotion(
                scopes,
                settings.session_promote_age_minutes,
                settings.session_promote_threshold,
            )
        )
        return list(result)


async def mark_promoted(scopes: Scopes, due: list[SessionItem]) -> None:
    """Stamp the promoted items so due_working_items never offers them again."""
    async with User.system(scopes) as session:
        await session.exec(
            update(SessionItem)
            .where(SessionItem.id.in_([item.id for item in due]))
            .values(promoted_at=func.now())
        )


async def promote_sessions(
    scopes: Scopes | None = None,
) -> int:
    """Feed a user's aged or overflow working items into the graph, return how many moved."""
    key = frozenset(scopes or (settings.system_user_id,))
    due = await due_working_items(key)
    if not due:
        return 0
    await ingest_texts(
        User.system(key),
        [
            TextSource(
                text=item.text,
                created_by=item.created_by,
                scopes=key,
                capture=CaptureContext.model_validate(item.provenance),
            )
            for item in due
        ],
    )
    await mark_promoted(key, due)
    await enqueue_pending(scopes=key)
    logger.info("promoted {} working items into graph scope {}", len(due), key)
    return len(due)

File: archive/ts/src/lib/server/graph/session_tier.ts
// The working-memory tier promotion, the TS twin of graph/session_tier.py: read one exact
// scope's still-working session items oldest first, pick the aged and overflow ones, feed
// them through the ingestion seam, and stamp promoted_at so they are never offered again.
// The TS server has no extract/ingest pipeline yet, so the graph ingestion is an injected
// seam the caller supplies; the Python side's enqueue_pending follow-up belongs to its
// pgqueuer worker and has no TS counterpart.
import { sql } from 'drizzle-orm';

import { settings } from '../settings';
import { asSystem, scopeKey, uuidArray } from './system';

export interface WorkingItem {
	id: string;
	created_at: Date;
	kind: string;
	text: string;
	created_by: string;
	provenance: Record<string, unknown>;
}

export type IngestTexts = (items: WorkingItem[], scopes: string[]) => Promise<void>;

// Return aged and overflow items in their existing oldest-first order, the
// SessionItem.due_for_promotion twin.
export const dueForPromotion = (
	items: WorkingItem[],
	now: Date,
	ageMinutes: number,
	threshold: number
): WorkingItem[] => {
	const cutoff = now.getTime() - ageMinutes * 60_000;
	const overflow = Math.max(0, items.length - threshold);
	return items.filter((item, index) => item.created_at.getTime() <= cutoff || index < overflow);
};

// Feed a user's aged or overflow working items into the graph, return how many moved, the
// promote_sessions twin with the ingestion pipeline injected.
export const promoteSessions = async (
	ingest: IngestTexts,
	scopes?: string[]
): Promise<number> => {
	const key = scopeKey(scopes);
	const now = new Date();
	const rows = await asSystem(key, async (tx) => {
		return (await tx.execute(sql`
			select id, created_at, kind, text, created_by, provenance from session_item
			where promoted_at is null and scopes = ${uuidArray(key)}::uuid[]
			order by created_at
		`)) as unknown as (Omit<WorkingItem, 'created_at'> & { created_at: string | Date })[];
	});
	// The driver hands timestamptz back as text here, so the age math gets real dates.
	const items = rows.map((row) => ({ ...row, created_at: new Date(row.created_at) }));
	const due = dueForPromotion(
		items,
		now,
		settings.sessionPromoteAgeMinutes,
		settings.sessionPromoteThreshold
	);
	if (!due.length) return 0;
	await ingest(due, key);
	await asSystem(key, (tx) =>
		tx.execute(sql`
			update session_item set promoted_at = ${now.toISOString()}::timestamptz
			where id = any(${uuidArray(due.map((item) => item.id))}::uuid[])
		`)
	);
	return due.length;
};
